main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -315,6 +315,31 @@ object BuildServerProtocol {
     bspBuildTargetCompileItem := bspCompileTask.value,
     bspBuildTargetRun := bspRunTask.evaluated,
     bspBuildTargetScalacOptionsItem := scalacOptionsTask.value,
+    bspBuildTargetJVMRunEnvironment := Def.inputTaskDyn {
+      val s = state.value
+      val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
+      val workspace = bspFullWorkspace.value.filter(targets)
+      val filter = ScopeFilter.in(workspace.scopes.values.toList)
+      Def.task {
+        val items = bspBuildTargetJvmEnvironmentItem.result.all(filter).value
+        val successfulItems = anyOrThrow(items)
+        val result = JvmRunEnvironmentResult(successfulItems.toVector, None)
+        s.respondEvent(result)
+      }
+    }.evaluated,
+    bspBuildTargetJVMTestEnvironment := Def.inputTaskDyn {
+      val s = state.value
+      val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
+      val workspace = bspFullWorkspace.value.filter(targets)
+      val filter = ScopeFilter.in(workspace.scopes.values.toList)
+      Def.task {
+        val items = bspBuildTargetJvmEnvironmentItem.result.all(filter).value
```

Comment:
These two tasks differ only here, some helper method might be introduced to reduce code repetition

Reply:
It also differ in the type of result: `JvmTestEnvironmentResult`. So I don't think it is worth doing anything here.

Reply:
These types are identical, I wonder if we need different types for both requests. The only reason I think of is being compliant with specification so, in the future, other people won't be confused why there is only one type for two endpoints.

Reply:
I'll leave different types for the sake of explicitness